chia/cmds/keys_funcs.py

```diff
@@ -107,28 +113,501 @@ def delete(fingerprint: int):
     Delete a key by its public key fingerprint (which is an integer).
     """
     print(f"Deleting private_key with fingerprint {fingerprint}")
-    keychain.delete_key_by_fingerprint(fingerprint)
+    Keychain().delete_key_by_fingerprint(fingerprint)
 
 
-@unlocks_keyring(use_passphrase_cache=True)
-def sign(message: str, fingerprint: int, hd_path: str, as_bytes: bool):
-    k = Keychain()
-    private_keys = k.get_all_private_keys()
+def derive_sk_from_hd_path(master_sk: PrivateKey, hd_path_root: str) -> Tuple[PrivateKey, str]:
+    """
+    Derive a private key from the provided HD path. Takes a master key and HD path as input,
+    and returns the derived key and the HD path that was used to derive it.
+    """
 
-    path: List[uint32] = [uint32(int(i)) for i in hd_path.split("/") if i != "m"]
-    for sk, _ in private_keys:
-        if sk.get_g1().get_fingerprint() == fingerprint:
-            for c in path:
-                sk = AugSchemeMPL.derive_child_sk(sk, c)
-            data = bytes.fromhex(message) if as_bytes else bytes(message, "utf-8")
-            print("Public key:", sk.get_g1())
-            print("Signature:", AugSchemeMPL.sign(sk, data))
-            return None
-    print(f"Fingerprint {fingerprint} not found in keychain")
+    from chia.wallet.derive_keys import _derive_path, _derive_path_unhardened
+
+    class DerivationType(Enum):
+        HARDENED = 0
```

Comment:
We should use Observer and non Observer (and probably change the `h` character to `no` in the HD paths). This was the decision that was made regarding terminology.

Reply:
Parsing the two letters of `no` might be a bit trickier, maybe just `n`? We can easily support multiple markers for hardened derivation. Where we see `if marker == HARDENED_CHAR:` just write `if marker in HARDENED_CHAR_SET:`. I'd recommend supporting both `n` and `h`, maybe `p` (and maybe `no` if it's not too difficult).

Reply:
Or maybe it's best to let the stupidly ambiguous `p` – the initial letter of both "public" and "private" – die the death it deserves.